### PR TITLE
Jasmine/Teaspoon testing infrastructure

### PR DESCRIPTION
**Pull request by ssayer**
Adds missing Rails Engine components and Jasmine/Teaspoon testing infrastructure.

Example spec included (but should be deleted eventually)

The massive number of files is mostly the dummy rails application that is used by Teaspoon to launch the testing code

### CODE REVIEW

File: lib/tasks/cql_qdm_patientapi_tasks.rake [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# desc "Explaining what the task does"
+# task :cql_qdm_patientapi do
+#   # Task goes here
+# end
```
Comment (hossenlopp) [inline lib/tasks/cql_qdm_patientapi_tasks.rake]: Why does this commented out file exist? Can it be deleted?

File: test/integration/navigation_test.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+require 'test_helper'
+
+class NavigationTest < ActionDispatch::IntegrationTest
+
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (hossenlopp) [inline test/integration/navigation_test.rb]: Does this file need to exist? I thought this was front end tests only?

File: test/integration/navigation_test.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+require 'test_helper'
```
Comment (mayerm94) [inline test/integration/navigation_test.rb]: Why do we need ruby tests? This repo is all front-end code I thought

File: test/teaspoon_env.rb [hunk 4]
```diff
@@ -0,0 +1,191 @@
+unless defined?(Rails)
+  ENV["RAILS_ROOT"] = File.expand_path("../dummy", __FILE__)
+  require File.expand_path("../dummy/config/environment", __FILE__)
+end
+
+Teaspoon.configure do |config|
```
Comment (hossenlopp) [inline test/teaspoon_env.rb]: This bombs out very fast here...

```
$ bundle exec teaspoon
/Users/hossenlopp/Documents/Bonnie/code/cql_qdm_patientapi/test/teaspoon_env.rb:6:in `<top (required)>': undefined method `configure' for Teaspoon:Module (NoMethodError)
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/environment.rb:47:in `load'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/environment.rb:47:in `require_env'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/environment.rb:20:in `require_environment'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/environment.rb:62:in `load_rails'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/environment.rb:6:in `load'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/console.rb:9:in `initialize'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/command_line.rb:15:in `new'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/lib/teaspoon/command_line.rb:15:in `initialize'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/bin/teaspoon:6:in `new'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/gems/teaspoon-1.1.5/bin/teaspoon:6:in `<top (required)>'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/bin/teaspoon:22:in `load'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/bin/teaspoon:22:in `<main>'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/bin/ruby_executable_hooks:15:in `eval'
	from /Users/hossenlopp/.rvm/gems/ruby-2.3.4/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by mayerm94** [comment]
I can't seem to get the tests to run in the browser while running a rails server, and I can't get a coverage report to generate. Do you have to do something special to get these working?

File: .travis.yml [hunk 5]
```diff
@@ -2,4 +2,15 @@ sudo: false
...
 rvm:
   - 2.2.2
-before_install: gem install bundler -v 1.12.3
+before_install:
+  - "sh -e /etc/init.d/xvfb start"
+before_install: gem install bundler -v 1.15.3
+script:
+  - export DISPLAY=:99.0
+  - bundle exec teaspoon
+  - bundle exec rake
```
Comment (hossenlopp) [inline .travis.yml]: I thought this was only frontend tests.

File: test/dummy/config/secrets.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+# Your secret key is used for verifying the integrity of signed cookies.
+# If you change this key, all old signed cookies will become invalid!
+
+# Make sure the secret is at least 30 characters and all random,
+# no regular words or you'll be exposed to dictionary attacks.
+# You can use `rake secret` to generate a secure secret key.
+
+# Make sure the secrets in this file are kept private
+# if you're sharing your code publicly.
```
Comment (ssayer) [inline test/dummy/config/secrets.yml]: I'm not sure what the issue here. This is in the "dummy" rails application that is used to mount the cql engine for testing purposes.

File: test/teaspoon_env.rb (hunk 4 above)
Comment (ssayer) [inline test/teaspoon_env.rb]: This should be fixed now

**Comment by hossenlopp** [comment]
This has my approval, but.... It needs to be tested with bonnie. Someone should point bonnie's gemfile at this branch and see if it breaks anything. I don't have time right now to do so.

File: test/dummy/config/secrets.yml (hunk 6 above)
Comment (c-monkey) [inline test/dummy/config/secrets.yml]: Fine by me

**Comment by c-monkey** [comment]
I'll test this now

**Comment by c-monkey** [comment]
Okay, this doesn't work:
```
MM197448-pc:bonnie ctohline$ bundle install
Fetching https://github.com/projectcypress/health-data-standards.git
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "rails":
  In snapshot (Gemfile.lock):
    rails (= 4.1.12)

  In Gemfile:
    rails (>= 4.0.0)

    cql_qdm_patientapi was resolved to 0.1.0, which depends on
      rails (~> 4.1.16)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
```

**Comment by ssayer** [comment]
Rails versions should be aligned now

**Comment by c-monkey** [comment]
Still failing, now due to coffee-rails. I suggest you pull Bonnie cql4bonnie branch, make this 1 change to the Gemfile (pointing to your local copy of the `testing` branch) and making sure `bundle install` works. Might as well run ruby and jasmine tests, to make sure nothing breaks

**Pull request merged**